### PR TITLE
PARQUET-285: Implement 3-level lists in Avro

### PR DESCRIPTION
**Pull request by rdblue**
This includes the write-side the changes from #83 that implement the 3-level list structure for parquet-avro. The old commit was https://github.com/rdblue/parquet-mr/commit/3589a7367c829b9eabc36b2e2e1cab31685415eb.

### CODE REVIEW

**Comment by rdblue** [comment]
@julienledem, this deprecates a method in ConversionPatterns so you might want to have a look at the first commit.

@SinghAsDev, can you do a first round of review?

**Comment by SinghAsDev** [comment]
@rdblue sure, adding to my todo list.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java [hunk 1]
```diff
@@ -425,4 +285,375 @@ private static GenericData getDataModel(Configuration conf) {
...
+        recordConsumer.startField("array", 0);
+        for (double element : array) {
+          recordConsumer.addDouble(element);
+        }
+        recordConsumer.endField("array", 0);
+      }
+    }
+  }
+
+  private class ThreeLevelWriter extends ListWriter {
```
Comment (SinghAsDev) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java]: Should we call it ThreeLevelListWriter?

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java [hunk 2]
```diff
@@ -425,4 +285,375 @@ private static GenericData getDataModel(Configuration conf) {
...
+
+    protected abstract void writeFloatArray(float[] array);
+
+    protected abstract void writeDoubleArray(double[] array);
+  }
+
+  /**
+   * For backward-compatibility. This preserves how lists were written in 1.x.
+   */
+  private class TwoLevelWriter extends ListWriter {
```
Comment (SinghAsDev) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java]: Should we call it TwoLevelListWriter?

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java [hunk 3]
```diff
@@ -105,6 +138,61 @@ public void testAllTypes() throws Exception {
...
+            "  required group mymap (MAP) {\n" +
+            "    repeated group map (MAP_KEY_VALUE) {\n" +
+            "      required binary key (UTF8);\n" +
+            "      required int32 value;\n" +
+            "    }\n" +
+            "  }\n" +
+            "  required group myemptymap (MAP) {\n" +
+            "    repeated group map (MAP_KEY_VALUE) {\n" +
+            "      required binary key (UTF8);\n" +
+            "      required int32 value;\n" +
```
Comment (SinghAsDev) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java]: I think we should add map with optional value as well.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java [hunk 4]
```diff
@@ -128,6 +216,9 @@ public void testAllTypes() throws Exception {
             "  optional group myoptionalarray (LIST) {\n" +
             "    repeated int32 array;\n" +
             "  }\n" +
+            "  required group myarrayofoptional (LIST) {\n" +
+            "    repeated int32 array;\n" +
```
Comment (SinghAsDev) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java]: myarrayofoptional with "repeated int32"? With repeated, elements will be optional?

**Comment by SinghAsDev** [comment]
Looks good to me, left a few minor comments and asks.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroSchemaConverter.java [hunk 5]
```diff
@@ -43,19 +45,23 @@
...
   public AvroSchemaConverter() {
     this.assumeRepeatedIsListElement = ADD_LIST_ELEMENT_RECORDS_DEFAULT;
+    this.writeOldListStructure = WRITE_OLD_LIST_STRUCTURE_DEFAULT;
   }
 
   public AvroSchemaConverter(Configuration conf) {
     this.assumeRepeatedIsListElement = conf.getBoolean(
         ADD_LIST_ELEMENT_RECORDS, ADD_LIST_ELEMENT_RECORDS_DEFAULT);
+    this.writeOldListStructure = conf.getBoolean(
+        WRITE_OLD_LIST_STRUCTURE, WRITE_OLD_LIST_STRUCTURE_DEFAULT);
```
Comment (julienledem) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroSchemaConverter.java]: the "reading from conf with default" should really be in the AvroReadSupport. The rest of the code should not depend on Hadoop conf

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java [hunk 6]
```diff
@@ -93,9 +104,19 @@ public WriteContext init(Configuration configuration) {
       this.rootAvroSchema = new Schema.Parser().parse(configuration.get(AVRO_SCHEMA));
       this.rootSchema = new AvroSchemaConverter().convert(rootAvroSchema);
     }
+
     if (model == null) {
       this.model = getDataModel(configuration);
     }
+
+    boolean writeOldListStructure = configuration.getBoolean(
+        WRITE_OLD_LIST_STRUCTURE, WRITE_OLD_LIST_STRUCTURE_DEFAULT);
```
Comment (julienledem) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java]: this line is duplicated

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java [hunk 7]
```diff
@@ -425,4 +285,375 @@ private static GenericData getDataModel(Configuration conf) {
...
+        recordConsumer.endField(LIST_ELEMENT_NAME, 0);
+        recordConsumer.endGroup();
+        recordConsumer.endField("array", 0);
+      }
+    }
+
+    @Override
+    protected void writeLongArray(long[] array) {
+      if (array.length > 0) {
+        recordConsumer.startField("array", 0);
```
Comment (julienledem) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java]: array did not get its constant?

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java [hunk 8]
```diff
@@ -425,4 +285,375 @@ private static GenericData getDataModel(Configuration conf) {
...
+        recordConsumer.endField("array", 0);
+      }
+    }
+
+    @Override
+    protected void writeFloatArray(float[] array) {
+      if (array.length > 0) {
+        recordConsumer.startField("array", 0);
+        recordConsumer.startGroup(); // repeated group array, middle layer
+        recordConsumer.startField(LIST_ELEMENT_NAME, 0);
```
Comment (julienledem) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java]: you could facto out some of this in a private method:
startArray() {
recordConsumer.startField("array", 0);
recordConsumer.startGroup(); // repeated group array, middle layer
recordConsumer.startField(LIST_ELEMENT_NAME, 0);
}
endArray() {
recordConsumer.endField(LIST_ELEMENT_NAME, 0);
recordConsumer.endGroup();
recordConsumer.endField("array", 0);
}

File: parquet-column/src/main/java/org/apache/parquet/schema/ConversionPatterns.java [hunk 9]
```diff
@@ -101,4 +107,26 @@ public static GroupType listType(Repetition repetition, String alias, Type neste
...
+    Preconditions.checkArgument(elementType.getName().equals(ELEMENT_NAME),
+        "List element type must be named 'element'");
+    return listWrapper(
+        listRepetition,
+        name,
+        LIST,
+        new GroupType(Repetition.REPEATED, "list", elementType)
+    );
+  }
 }
```
Comment (julienledem) [inline parquet-column/src/main/java/org/apache/parquet/schema/ConversionPatterns.java]: looks good to me

**Comment by julienledem** [comment]
I've made some comments. Otherwise this looks good to me.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java (hunk 4 above)
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java]: No, this is testing the old behavior and verifying that there is no change when the configuration options to turn on the new behavior are not set. Basically validating backward-compatibility for code.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java (hunk 3 above)
Comment (rdblue) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java]: I'm fine with that, but how about as a separate issue? This doesn't change the structure of maps.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java (hunk 2 above)
Comment (rdblue) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java]: Done.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java (hunk 1 above)
Comment (rdblue) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java]: Done.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java (hunk 8 above)
Comment (rdblue) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java]: Good point, then I can combine all of the type-specific methods on the ListWriter class, too.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java (hunk 7 above)
Comment (rdblue) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroWriteSupport.java]: This shouldn't even be array. it should be LIST_REPEATED_NAME.

**Comment by rdblue** [comment]
Fixed most of the issues, but I still need to take a look at how configuration is done.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java (hunk 3 above)
Comment (SinghAsDev) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java]: I am fine with it.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java (hunk 4 above)
Comment (SinghAsDev) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java]: Ok, makes sense now.

**Comment by rdblue** [comment]
@julienledem, I think the configuration changes are going to be fairly significant. I'd like to get this in and update how parquet-avro manages configuration in a separate issue. Does that sound reasonable?

**Comment by julienledem** [comment]
@rdblue agreed. Since the existing code is already using conf in that way, improvements can be in a follow up.

**Comment by rdblue** [comment]
@julienledem, okay to merge this, then?

**Comment by julienledem** [comment]
+1

**Comment by rdblue** [comment]
Thanks for the reviews @SinghAsDev and @julienledem!

**Pull request closed without merging**